adapters/clydo/clydo.go

```diff
@@ -0,0 +1,294 @@
+	if request.Device == nil {
+		return nil, &errortypes.BadInput{Message: "Failed to get device headers"}
+	}
+
+	var headers []struct{ key, value string }
+
+	if ipv6 := request.Device.IPv6; ipv6 != "" {
+		headers = append(headers, struct{ key, value string }{"X-Forwarded-For", ipv6})
+	}
+	if ip := request.Device.IP; ip != "" {
+		headers = append(headers, struct{ key, value string }{"X-Forwarded-For", ip})
+	}
+	if ua := request.Device.UA; ua != "" {
+		headers = append(headers, struct{ key, value string }{"User-Agent", ua})
+	}
+
+	return headers, nil
+}
+
+func prepareImpIds(request *openrtb2.BidRequest) ([]string, error) {
+	impIds := openrtb_ext.GetImpIDs(request.Imp)
+	if impIds == nil {
+		return nil, &errortypes.BadInput{Message: "Failed to get imp ids"}
+	}
+	return impIds, nil
+}
+
+func checkResponseStatus(responseData *adapters.ResponseData) []error {
+	switch responseData.StatusCode {
+	case http.StatusNoContent:
+		return []error{}
+	case http.StatusBadRequest:
+		return []error{&errortypes.BadInput{
+			Message: "Bad request. Run with request.debug = 1 for more info.",
+		}}
+	case http.StatusOK:
+		return nil
+	default:
+		return []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}}
+	}
+}
+
+func prepareBidResponse(body []byte) (openrtb2.BidResponse, error) {
+	var response openrtb2.BidResponse
+	if err := jsonutil.Unmarshal(body, &response); err != nil {
+		return response, err
+	}
+	return response, nil
+}
+
+func prepareSeatBids(seatBids []openrtb2.SeatBid, bidTypeMap map[string]openrtb_ext.BidType) ([]*adapters.TypedBid, []error) {
+	var typedBids []*adapters.TypedBid
+	var errors []error
+
+	if seatBids == nil {
+		return typedBids, nil
+	}
+
+	for _, seatBid := range seatBids {
+		if seatBid.Bid == nil {
+			continue
+		}
+		for i := range seatBid.Bid {
+			bid := &seatBid.Bid[i]
+			bidType := getMediaTypeForBid(bid, bidTypeMap)
+			fmt.Println(bidType)
+			typedBids = append(typedBids, &adapters.TypedBid{
+				Bid:     bid,
+				BidType: bidType,
+			})
+		}
+	}
+
+	return typedBids, errors
+}
+
+func buildBidTypeMap(imps []openrtb2.Imp) (map[string]openrtb_ext.BidType, error) {
+	bidTypeMap := make(map[string]openrtb_ext.BidType, len(imps))
+	for _, imp := range imps {
+		switch {
+		case imp.Video != nil:
+			bidTypeMap[imp.ID] = openrtb_ext.BidTypeVideo
+		case imp.Native != nil:
+			bidTypeMap[imp.ID] = openrtb_ext.BidTypeNative
+		case imp.Banner != nil:
+			bidTypeMap[imp.ID] = openrtb_ext.BidTypeBanner
+		default:
+			return nil, &errortypes.BadInput{
+				Message: "Failed to get media type",
+			}
+		}
+	}
+	return bidTypeMap, nil
+}
+
+func getMediaTypeForBid(bid *openrtb2.Bid, bidTypeMap map[string]openrtb_ext.BidType) openrtb_ext.BidType {
+	// if mediaType, ok := bidTypeMap[bid.ImpID]; ok {
+	// 	return mediaType, nil
+	// }
```

Comment:
Those commented lines seemed like a good idea in order to cover the edge case where `bid.ImpID` isn't found in `bidTypeMap`. Can they get uncommented? If not, can they get removed?